### PR TITLE
Modernizing cmake

### PR DESCRIPTION
**Pull request by beckerhe**
Brace yourself for a large pull request.

The idea behind this work is to standardize our build system to make it work on CI more easily.
Recent versions of CMake brought a lot of new features which radically decrease the complexity
of cmake based build systems.

The following points have been tackled:
- Properly separate external dependencies form internal libraries/modules.
- Separation between compiler toolchain flags and "normal" compiler flags by introducing toolchain files.
- Proper handling of private and public dependencies (this mainly applies to include directories, compiler flags and linking)

# Dependencies
Dependencies in CMake are usually imported by a find_package-call. The result of find_package-call in modern CMake is an imported target (an enitity which bundles link libraries, include paths, compiler flags, etc.). A module in our project can refer to such an imported target by leveraging the target_link_libraries()-command.

All 3rd party dependencies in ORBIT are now handled by a find_package-call and subsequent imports with target_link_libraries. For all dependencies which do not provide CMake support a FindXYZ.cmake package finding file was written (check the cmake/ folder).

Furthermore, both on Windows and on Linux, all dependencies (except Qt) are pulled form vcpkg. For making CI happen, we should work towards having all dependencies in vcpkg and cross-compiling form source against the GGP SDK (at least for the headless part).

# Toolchain files
Like I mentioned before, I removed all compiler flags from the build system which are not directly necessary for building the project (like -D options). All compiler flags (like optimization levels, enabled and disabled warnings, LTO, etc.) should be defined in a toolchain file. This way, concerns are separated and we make sure to compile the whole project with a unified set of compiler flags.

The are several example toolchain files for both Windows and Linux in contrib/toolchains. Feel free to use them as a start.

# Deployment & Packaging
There has already been some cmake code which made a deployment possible by collecting all the required libraries. I believe this was not completely up to date, but close. For now I removed theses parts (my apologies), just to reduce the scope of this pull request as much as possible. If you think that's a problem, please let me know. The idea is to bring this back soon. We will need that anyway.

# CMake formatting
For alle the CMake files I used cmake-format (python pip package) to achieve a consistent style. The formatting is sometimes not perfect, but at least I don't need to worry about that.

# Testing
I tried on Windows x64 debug and release and on Linux x64 debug and release. Daniel is currently checking the bootstraping on his windows machine.

# CMake version requirements
Yes, I had to bump up the cmake version to a pretty recent version (3.15). The reason is, that CMake made quite some progress in the last two years and brought features which greatly simplify the CMakeLists.txt-files and allow a cleaner build configuration. I know, this complicates things, especially on Linux, where you usually use CMake from your package manager. But I still believe it is worth the hassle. If you don't agree, please let me know. (By the way, the easiest way to get a recent version of CMake on Linux is through pip: pip3 install --user cmake)

@danielfenner Daniel, you wanted to make the code review for this. So be my guest. :-)
@pierricgimmig I believe there is no way forward without you having a look. So thank you in advance.

I hope, I haven't forgotten anything. In case, just let me know.

### CODE REVIEW

**Comment by danielfenner** [comment]
When running bootstrap-orbit.bat on Windows the build got hung up while building breakpad. I needed to press "enter" to have it continue (!?). This only happened during the x86 (not the x64) build. I didn't see a message indicating the root cause of this weird behaviour.

Finally the build failed due to some references to vs 2019 in some file. Let's have a look at this on my machine tomorrow.

**Comment by pierricgimmig** [comment]
@beckerhe , this looks fantastic, so much cleaner, I love it!

However, when running bootstrap-orbit.sh from scratch (after cloning your repo and switching to cleanup_cmake), I run into issues. 

The first one is vcpkg complaining about an error in freetype-gl/portfile.cmake.  The problem spawns from the removal of this block in the bootstrap script:

## Override freetype-gl portfile for linux (.lib->.a)
cd ../..
cp "OrbitUtils/freetype-gl-portfile.cmake" "external/vcpkg/ports/freetype-gl/portfile.cmake"
cd external/vcpkg

There is an error in the portfile, we should really check if it is fixed in the vcpkg repor, or submit a fix, but in the meantime I think we should re-add the block in the script. 

I also get linking errors:
undefined reference to `hb_font_destroy'
undefined reference to `hb_ot_tags_from_script'
undefined reference to `hb_set_create'
undefined reference to `hb_ot_layout_collect_lookups'
undefined reference to `hb_set_is_empty'
undefined reference to `hb_set_create'
undefined reference to `hb_ot_layout_lookup_collect_glyphs'
undefined reference to `hb_set_next'
...

I'm guessing you didn't see those errors on your side because you had your vcpkg directory already populated from a previous initialization.  Could you clone your repo and bootstrap from scratch to see if you run into the same issues?

**Comment by beckerhe** [comment]
> @beckerhe , this looks fantastic, so much cleaner, I love it!
> 
> However, when running bootstrap-orbit.sh from scratch (after cloning your repo and switching to cleanup_cmake), I run into issues.
> 
> The first one is vcpkg complaining about an error in freetype-gl/portfile.cmake. The problem spawns from the removal of this block in the bootstrap script:
> 
> ## Override freetype-gl portfile for linux (.lib->.a)
> cd ../..
> cp "OrbitUtils/freetype-gl-portfile.cmake" "external/vcpkg/ports/freetype-gl/portfile.cmake"
> cd external/vcpkg
> 
> There is an error in the portfile, we should really check if it is fixed in the vcpkg repor, or submit a fix, but in the meantime I think we should re-add the block in the script.

You're absolutely right. I wanted to check whether there is any cleaner way, but I forgot about it.
And there isn't a better way (apart from updating vcpkg, which should be done in a separate step.)
> 
> I also get linking errors:
> undefined reference to `hb_font_destroy' undefined reference to `hb_ot_tags_from_script'
> undefined reference to `hb_set_create' undefined reference to `hb_ot_layout_collect_lookups'
> undefined reference to `hb_set_is_empty' undefined reference to `hb_set_create'
> undefined reference to `hb_ot_layout_lookup_collect_glyphs' undefined reference to `hb_set_next'
> ...
> 
> I'm guessing you didn't see those errors on your side because you had your vcpkg directory already populated from a previous initialization. Could you clone your repo and bootstrap from scratch to see if you run into the same issues?

I just made a clean bootstrap and I don't see these linking errors. These symbols seem to come from HarfBuzz which is a font library like freetype and it might be the case that freetype links against HarfBuzz. But that's not the case on my machine. And it shouldn't be on your machine either because freetype in bootstrap is taken from vcpkg. Can you check which object file or library needs these symbols? Somthing like `find build/ external/vcpkg/installed/x64-linux/lib/ -name \*.a -or -name \*.o | xargs nm | grep hb_`?

**Comment by beckerhe** [comment]
I just merged headless (with all the clang-format changes) into this branch.

File: bootstrap-orbit.sh [hunk 1]
```diff
@@ -26,9 +26,12 @@ fi
...
-mkdir build
+  mkdir build
 fi
+
 cd build
-cmake -DCMAKE_BUILD_TYPE=Debug ..
-#cmake ..
-make
+if [ ! -f toolchain.cmake ]; then
+  cp ../contrib/toolchains/toolchain-linux-default-release.cmake toolchain.cmake
```
Comment (dimitry-) [inline bootstrap-orbit.sh]: ls -s, also use clang-toolchain by default?

**Comment by beckerhe** [comment]
@danielfenner 
Hi Daniel,
bootstrapping on Windows now works smoothly, at least on my machine.

I also adjusted the bootstrapping script to use VS2017 by default. So your changes shouldn't be necessary anymore. Can you try once more and report back whether it compiles?

**Comment by pierricgimmig** [comment]
> I just made a clean bootstrap and I don't see these linking errors. These symbols seem to come from HarfBuzz which is a font library like freetype and it might be the case that freetype links against HarfBuzz.

This is weird, I don't understand then.  I get: 

/usr/bin/ld: ../external/vcpkg/installed/x64-linux/lib/libfreetype.a(autofit.c.o): in function `af_face_globals_free':
autofit.c:(.text+0x2b95): undefined reference to `hb_font_destroy'

Looking into freetype's source, in "external/vcpkg/buildtrees/freetype/src/2.9.1-7d839e348f/src/autofit/afglobal.c" we have:

#ifdef FT_CONFIG_OPTION_USE_HARFBUZZ
      hb_font_destroy( globals->hb_font );
      hb_buffer_destroy( globals->hb_buf );
#endif

FT_CONFIG_OPTION_USE_HARFBUZZ is defined in external/vcpkg/buildtrees/freetype/src/2.9.1-7d839e348f/devel/ftoption.h.  

For this to compile on your side you either have FT_CONFIG_OPTION_USE_HARFBUZZ *not* defined, or you are actually linking with harfbuzz somehow?  Maybe you have a newer version of vcpkg that addressed the issue?

**Comment by pierricgimmig** [comment]
Ok so I have a bit more information, I have harfbuzz installed on my machine, the headers are found in "/usr/include/harfbuzz".  Because freetype found harfbuzz it enables FT_CONFIG_OPTION_USE_HARFBUZZ in the code.  A workaround was to remove harfbuzz from my system, then the harfbuzz related linking error disappeared.  I have the feeling this might have been fixed in a newer versions of vcpkg/freetype, so I'll wait until we update vcpkg instead of fixing the issue now.

Also, I ran into another linking error related to the fact that we are not linking with *freeglut* anymore but with *glut*.  We should really re-enable *freeglut*.  From https://en.wikipedia.org/wiki/FreeGLUT:   "Since GLUT has gone into stagnation, FreeGLUT is in development to improve the toolkit."  See attached [patch](https://github.com/pierricgimmig/orbitprofiler/files/4131488/freeglut_patch.txt) for the fix on Linux (not tested on Windows).

**Comment by beckerhe** [comment]
> Ok so I have a bit more information, I have harfbuzz installed on my machine, the headers are found in "/usr/include/harfbuzz". Because freetype found harfbuzz it enables FT_CONFIG_OPTION_USE_HARFBUZZ in the code. A workaround was to remove harfbuzz from my system, then the harfbuzz related linking error disappeared. I have the feeling this might have been fixed in a newer versions of vcpkg/freetype, so I'll wait until we update vcpkg instead of fixing the issue now.
Hi Pierric,
I checked the ports file in vcpkg's master branch. Your feeling was right, it is fixed there. You can choose in future versions whether you want to build with or without harfbuzz. I expect the system harfbuzz not to be picked up if you choose not to build with harfbuzz support. But I haven't tried yet.
> 
> Also, I ran into another linking error related to the fact that we are not linking with _freeglut_ anymore but with _glut_. We should really re-enable _freeglut_. From https://en.wikipedia.org/wiki/FreeGLUT: "Since GLUT has gone into stagnation, FreeGLUT is in development to improve the toolkit." See attached [patch](https://github.com/pierricgimmig/orbitprofiler/files/4131488/freeglut_patch.txt) for the fix on Linux (not tested on Windows).
Actually we should still be linking against freeglut, even if it seems not to be the case. CMake's `FindGLUT.cmake`-module supports all different flavors of GLUT, including freeglut. On my system it reliably picks up the `libglut.a` from the vcpkg `installed/`-directory. Can you check if that's also the case for you? `grep GLUT_glut_LIBRARY CMakeCache.txt` should tell you the content of `GLUT_glut_LIBRARY`.

One more thing: vcpkg compiles freeglut with the option `FREEGLUT_REPLACE_GLUT` by default on Linux. That's why the library is called `libglut.a` instead of `libfreeglut.a`, even if it originates from the freeglut package.

If all of this looks good, can you share the linking errors?

Cheers,
Henning

**Comment by antonrohr** [comment]
Hi, 

I just did a fresh clone and bootstrapped under windows (x64). It worked almost perfectly, only glew x86 build failed, which is not needed on my x64 system. 
`Error: Building package glew:x86-windows failed with: BUILD_FAILED` . The next time I ran it, glew compiled successfully. (This might be connected to some malware detection service blocking some file access or something similar)

To actually run Orbit.exe you still need to run `windeployqt.exe .\Orbit.exe`.

Also, there is no Orbit.sln (Visual Studio Solution) in the build folder anymore, but a peparse.sln, which contains all Orbit targets plus some other targets. (This is already fixed by Henning in commit dc584dd)

All targets in the solution build in VS without problems under Release and Debug, but under Debug OrbitQT does not start. There are some debug libraries missing. If I copy them from `\external\vcpkg\installed\x64-windows\debug\bin` and rename them appropriately, Orbit runs and I can use breakpoints to debug. I spoke to Henning in person and he thinks its a vcpkg problem, that might be fixed in his branch [update_vcpkg](/beckerhe/orbitprofiler/tree/update_vcpkg)

Best
Anton

**Comment by pierricgimmig** [comment]
> Can you check if that's also the case for you? `grep GLUT_glut_LIBRARY CMakeCache.txt` should tell you the content of `GLUT_glut_LIBRARY`.

Grepping gives: 

GLUT_glut_LIBRARY:FILEPATH=orbitprofiler/external/vcpkg/installed/x64-linux/lib/libglut.a
//ADVANCED property for variable: GLUT_glut_LIBRARY
GLUT_glut_LIBRARY-ADVANCED:INTERNAL=1

Building your branch as is I get these linking error:

/usr/bin/ld: ../external/vcpkg/installed/x64-linux/lib/libglut.a(fg_gamemode_x11.c.o): in function `xrandr_resize':
fg_gamemode_x11.c:(.text+0x2f): undefined reference to `XRRQueryVersion'
/usr/bin/ld: fg_gamemode_x11.c:(.text+0x8a): undefined reference to `XRRGetScreenInfo'
/usr/bin/ld: fg_gamemode_x11.c:(.text+0xa3): undefined reference to `XRRConfigSizes'
/usr/bin/ld: fg_gamemode_x11.c:(.text+0xb5): undefined reference to `XRRConfigCurrentConfiguration'
/usr/bin/ld: fg_gamemode_x11.c:(.text+0xcb): undefined reference to `XRRConfigTimes'
/usr/bin/ld: fg_gamemode_x11.c:(.text+0x15d): undefined reference to `XRRSetScreenConfig'

**Comment by beckerhe** [comment]
> > Can you check if that's also the case for you? `grep GLUT_glut_LIBRARY CMakeCache.txt` should tell you the content of `GLUT_glut_LIBRARY`.
> 
> Grepping gives:
> 
> GLUT_glut_LIBRARY:FILEPATH=orbitprofiler/external/vcpkg/installed/x64-linux/lib/libglut.a
> //ADVANCED property for variable: GLUT_glut_LIBRARY
> GLUT_glut_LIBRARY-ADVANCED:INTERNAL=1

At least that looks good. It's not trying to link against the system-wide library.
> 
> Building your branch as is I get these linking error:
> 
> /usr/bin/ld: ../external/vcpkg/installed/x64-linux/lib/libglut.a(fg_gamemode_x11.c.o): in function `xrandr_resize': fg_gamemode_x11.c:(.text+0x2f): undefined reference to `XRRQueryVersion'
> /usr/bin/ld: fg_gamemode_x11.c:(.text+0x8a): undefined reference to `XRRGetScreenInfo' /usr/bin/ld: fg_gamemode_x11.c:(.text+0xa3): undefined reference to `XRRConfigSizes'
> /usr/bin/ld: fg_gamemode_x11.c:(.text+0xb5): undefined reference to `XRRConfigCurrentConfiguration' /usr/bin/ld: fg_gamemode_x11.c:(.text+0xcb): undefined reference to `XRRConfigTimes'
> /usr/bin/ld: fg_gamemode_x11.c:(.text+0x15d): undefined reference to `XRRSetScreenConfig'

These are XRandR symbols, aren't they? Can you try adding the target X11::Xrandr to OrbitGl's lib dependencies? (OrbitGl/CMakeLists.xtxt, line 112)

**Comment by pierricgimmig** [comment]
> Can you try adding the target X11::Xrandr to OrbitGl's lib dependencies?

Totally fixes the issue!  Thanks!

**Comment by beckerhe** [comment]
> > Can you try adding the target X11::Xrandr to OrbitGl's lib dependencies?
> 
> Totally fixes the issue! Thanks!

That's weird. Probably a bug in vcpkg's freeglut package. My guess is that it's picking
up some xrandr-headers which haven been installed on your system. Uninstalling libxrandr-dev might fix that, even if it's more a workaround than a solution.

I just had a quick look in freeglut's build configuration and it's - let's say - every package maintainer's best friend. Tons of optional dependencies. And xrandr is one of them. Here is a snippet:
```cmake
FIND_PACKAGE(X11 REQUIRED)
    INCLUDE_DIRECTORIES(${X11_X11_INCLUDE_PATH})
    LIST(APPEND LIBS ${X11_X11_LIB})
    IF(X11_Xrandr_FOUND)
        SET(HAVE_X11_EXTENSIONS_XRANDR_H TRUE)
        LIST(APPEND LIBS ${X11_Xrandr_LIB})
    ENDIF()
```

I'm going to check tomorrow, whether newer versions of vcpkg do something about randomly picking up optional dependencies.

**Comment by beckerhe** [comment]
Hi Pierric,
I believe this PR is now in such a state that it can be merged.

There are still several issues, which are related to bugs in vcpkg (like your xrandr bug).
These can be fixed by updating vcpkg, but I think this shouldn't be part of this merge request.

What do you think?

Henning

**Comment by pierricgimmig** [comment]
> What do you think?

Hi Henning, do you think we can include the X11::Xrandr fix as part of this PR in case others run into the same issue?

**Comment by beckerhe** [comment]
> > What do you think?
> 
> Hi Henning, do you think we can include the X11::Xrandr fix as part of this PR in case others run into the same issue?

Sure. Sorry, I should have made myself more clear. My point is that we have to live with workarounds in this branch and then fix it in a later stage when updating vcpkg.

I'm going to push the fix in a couple of minutes.

Cheers,
Henning

**Comment by pierricgimmig** [comment]
Note: on my Ubuntu 19.10, 
- I had to manually install ninja (sudo apt-get install ninja-build).  Should we add "ninja-build" to bootstrap-orbit.sh?
- harfbuzz was installed so I ran into the same linking issue as above.  Had to rename /usr/include/harfbuzz/ to /usr/include/harfbuzz_/ temporarily so that vcpkg wouldn't pick it up.
- I had to add "X11::Xrandr" to the list of dependencies in OrbitGl/CMakeLists (line 112) to fix a linking error.
- Orbit crashes on startup (I'm investigating now...)

Still, I can still merge the PR (once you add the fix) in hope that it will unblock you for the following changes that will upgrade vcpkg and possibly fix those issues.

**Comment by pierricgimmig** [comment]
@beckerhe , I'm merging the CL even if the branch won't be in perfect shape.

By the way, CLion used to be able to build Orbit out of the box, however I'm running issues now:

"CMake Error at CMakeLists.txt:22 (find_package):
  By not providing "Findcereal.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "cereal", but
  CMake did not find one."

It would be great to be able to build from Clion again.

**Comment by beckerhe** [comment]
> @beckerhe , I'm merging the CL even if the branch won't be in perfect shape.
> 
> By the way, CLion used to be able to build Orbit out of the box, however I'm running issues now:
> 
> "CMake Error at CMakeLists.txt:22 (find_package):
> By not providing "Findcereal.cmake" in CMAKE_MODULE_PATH this project has
> asked CMake to find a package configuration file provided by "cereal", but
> CMake did not find one."
> 
> It would be great to be able to build from Clion again.

That's probably because it's not loading the vcpkg-toolchain file by default anymore. I'm thinking about adding this back. CLion is actually a viable use case for this. Expect a pull request today or on Monday. :-)

Thanks for merging,
Henning

**Pull request merged**